3p/ampcontext-integration.js

```diff
@@ -0,0 +1,153 @@
+    win.name = masterName;
+    master = win;
+  }
+  return master;
+}
+
+
+export class IntegrationAmpContext extends AbstractAmpContext {
+
+  /** @param {!Window} win */
+  constructor(win) {
+    super(win);
+
+    /** @private {?string} */
+    this.embedType_ = null;
+
+    /** @private {boolean} */
+    this.updateDimensionsEnabled_ = false;
+  }
+
+  /** @override */
+  isAbstractImplementation_() {
+    return false;
+  }
+
+  /** @return {!Window} */
+  get master() {
+    return this.master_();
+  }
+
+  /** @return {!Window} */
+  master_() {
+    return masterSelection(this.win_, dev().assertString(this.embedType_));
+  }
+
+  /** @return {boolean} */
+  get isMaster() {
+    return this.isMaster_();
+  }
+
+  /** @return {boolean} */
+  isMaster_() {
+    return this.master == this.win_;
+  }
+
+  /** @param {string} type */
+  setEmbedType(type) {
+    this.embedType_ = type;
+
+    if (type === 'facebook' || type === 'twitter') {
+      // Only make this available to selected embeds until the
+      // generic solution is available.
+      this.updateDimensionsEnabled_ = true;
+    }
+  }
+
+  /**
+   * @param {number} width
+   * @param {number} height
+   */
+  updateDimensions(width, height) {
+    dev().assert(this.updateDimensionsEnabled_, 'Not available.');
+    this.requestResize(width, height);
+  }
+
+  notifyBootstrapLoaded() {
+    this.client_.sendMessage('bootstrap-loaded');
+  }
+
+  /**
+   * @param {{width, height}=} opt_data
+   */
+  renderStart(opt_data) {
+    this.client_.sendMessage('render-start', opt_data);
+  }
+
+  /**
+   * Reports the "entity" that was rendered to this frame to the parent for
+   * reporting purposes.
+   * The entityId MUST NOT contain user data or personal identifiable
+   * information. One example for an acceptable data item would be the
+   * creative id of an ad, while the user's location would not be
+   * acceptable.
+   * @param {string} entityId See comment above for content.
+   */
+  // DUPLICATE IN 3p/integration
+  reportRenderedEntityIdentifier(entityId) {
+    this.client_.sendMessage('entity-id', {
+      id: user().assertString(entityId),
+    });
+  }
+
+  /**
+   * Performs a potentially asynchronous task exactly once for all frames of a
+   * given type and the provide the respective value to all frames.
+   * @param {!Window} global Your window
+   * @param {string} taskId Must be not conflict with any other global variable
+   *     you use. Must be the same for all callers from all frames that want
+   *     the same result.
+   * @param {function(function(*))} work Function implementing the work that
+   *     is to be done. Receives a second function that should be called with
+   *     the result when the work is done.
+   * @param {function(*)} cb Callback function that is called when the work is
+   *     done. The first argument is the result.
+   */
+  computeInMasterFrame(global, taskId, work, cb) {
+    computeInMasterFrame(global, taskId, work, cb);
```

Comment:
It's being used in other modules as well. I'd be happy to test for race conditions caused by making all use `window.context.computeInMasterFrame` separately.